src/libraries/System.IO.Compression.Brotli/src/System/IO/Compression/dec/BrotliStream.Decompress.cs

```diff
@@ -57,59 +57,63 @@ public override int Read(Span<byte> buffer)
+                    _bufferCount -= bytesConsumed;
+                }
+
+                // If we successfully decompressed any bytes, or if we've reached the end of the decompression, we're done.
+                if (bytesWritten > 0 || lastResult == OperationStatus.Done)
+                {
+                    break;
+                }
+
+                // If we need more input data, read from the underlying stream.
                 if (lastResult == OperationStatus.NeedMoreData)
                 {
                     // Ensure any left over data is at the beginning of the array so we can fill the remainder.
                     if (_bufferCount > 0 && _bufferOffset != 0)
                     {
-                        _buffer.AsSpan(_bufferOffset, _bufferCount).CopyTo(_buffer);
+                        new ReadOnlySpan<byte>(_buffer, _bufferOffset, _bufferCount).CopyTo(_buffer);
                     }
                     _bufferOffset = 0;
 
-                    int numRead = 0;
-                    while (_bufferCount < _buffer.Length && ((numRead = _stream.Read(_buffer, _bufferCount, _buffer.Length - _bufferCount)) > 0))
-                    {
-                        _bufferCount += numRead;
-                        if (_bufferCount > _buffer.Length)
-                        {
-                            // The stream is either malicious or poorly implemented and returned a number of
-                            // bytes larger than the buffer supplied to it.
-                            throw new InvalidDataException(SR.BrotliStream_Decompress_InvalidStream);
-                        }
-                    }
-
-                    if (_bufferCount <= 0)
+                    int numRead = _stream.Read(_buffer, _bufferCount, _buffer.Length - _bufferCount);
+                    if (numRead <= 0)
                     {
                         break;
                     }
-                }
 
-                lastResult = _decoder.Decompress(new ReadOnlySpan<byte>(_buffer, _bufferOffset, _bufferCount), buffer, out int bytesConsumed, out int bytesWritten);
-                if (lastResult == OperationStatus.InvalidData)
-                {
-                    throw new InvalidOperationException(SR.BrotliStream_Decompress_InvalidData);
-                }
+                    _bufferCount += numRead;
 
-                if (bytesConsumed > 0)
-                {
-                    _bufferOffset += bytesConsumed;
-                    _bufferCount -= bytesConsumed;
+                    if (_bufferCount > _buffer.Length)
+                    {
+                        ThrowInvalidStream();
+                    }
                 }
 
-                if (bytesWritten > 0)
+                // Now that we know data is available, if our input buffer is zero-byte, we're done.
+                // We wait to do this check until after reading in order to support zero-byte reads
+                // waiting for data.
+                if (buffer.IsEmpty)
```

Comment:
Technically you don't know that data is available. You could have read an amount insufficient to decode a character. To ensure you can decode at least one character, it seems like you'd need to loop back to the decoder and see whether you get NeedMoreData again.

Reply:
Also, is the decoder here guaranteed to prefer returning NeedsMoreData over the result that means "I filled your whole buffer so I stopped decoding" when you pass it a 0-byte buffer? It must, or the tests would pass, right?

Reply:
> Technically you don't know that data is available.

I was more descriptive about that for the same situation in DeclateStream... I can elaborate here as well. 

> Also, is the decoder here guaranteed to prefer returning NeedsMoreData

That's a good question.  I'll look at it tomorrow.  I suspect not.